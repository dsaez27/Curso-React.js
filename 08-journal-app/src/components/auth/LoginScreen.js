import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { useForm } from './../../hooks/useForm';
import {
    startGoogleLogin,
    startLoginEmailPassword,
} from '../../redux/actions/auth';

export const LoginScreen = () => {
    const dispatch = useDispatch();
    const { loading } = useSelector((state) => state.ui);

    const [formValues, handleInputChange] = useForm({
        email: 'Daniel',
        password: '123456',
    });

    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        isFormValid && dispatch(startLoginEmailPassword(email, password));
    };

    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin());
    };

    const isFormValid = () => {
        if (email.trim().length === 0) {
            return false;
        } else if (password.trim().length === 0) {
            return false;
        }

        return true;
    };

    return (
        <>
            <h3 className='auth__title'>Login</h3>

            <form onSubmit={handleLogin}>
                <input
                    className='auth__input'
                    type='text'
                    placeholder='Email'
                    name='email'
                    autoComplete='off'
                    value={email}
                    onChange={handleInputChange}
                />
                <input
                    className='auth__input'
                    type='password'
                    placeholder='Password'
                    name='password'
                    autoComplete='off'
                    value={password}
                    onChange={handleInputChange}
                />
                <button
                    className='btn btn-primary btn-block'
                    type='submit'
                    disabled={loading}
                >
                    Login
                </button>
                <div className='auth__social-networks'>
                    <p>Login with social networks</p>
                    <div className='google-btn' onClick={handleGoogleLogin}>
                        <div className='google-icon-wrapper'>
                            <img
                                className='google-icon'
                                src='https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg'
                                alt='google button'
                            />
                        </div>
                        <p className='btn-text'>
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>
                <Link to='auth/register' className='link'>
                    Create new account
                </Link>
            </form>
        </>
    );
};
