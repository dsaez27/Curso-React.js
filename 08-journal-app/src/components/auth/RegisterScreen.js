import React from 'react';
import { Link } from 'react-router-dom';

export const RegisterScreen = () => {
    return (
        <>
            <h3 className='auth__title'>Register</h3>

            <form>
                <input
                    className='auth__input mb-1'
                    type='text'
                    placeholder='Name'
                    name='name'
                    autoComplete='off'
                />
                <input
                    className='auth__input mb-1'
                    type='text'
                    placeholder='Email'
                    name='email'
                    autoComplete='off'
                />
                <input
                    className='auth__input mb-1'
                    type='password'
                    placeholder='Password'
                    name='password'
                />
                <input
                    className='auth__input mb-1'
                    type='password'
                    placeholder='Confirm Password'
                    name='password2'
                />
                <button
                    className='btn btn-primary btn-block mb-5'
                    type='submit'
                >
                    Register
                </button>

                <Link to='auth/login' className='link'>
                    Already registered?
                </Link>
            </form>
        </>
    );
};
