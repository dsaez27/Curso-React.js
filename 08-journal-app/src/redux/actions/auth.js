import Swal from 'sweetalert2';
import { types } from '../types';
import { finishLoading, setError, startLoading } from './ui';
import { noteLogout } from './notes';
import { firebase, googleAuthProvider } from '../../firebase/firebaseConfig';

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch(startLoading());
        try {
            return firebase
                .auth()
                .signInWithEmailAndPassword(email, password)
                .then(({ user }) => {
                    dispatch(login(user.uid, user.displayName));
                    dispatch(finishLoading());
                })
                .catch((e) => {
                    dispatch(finishLoading());
                    Swal.fire({
                        title: 'Error',
                        text: e.message,
                        icon: 'error',
                    });
                });
        } catch (e) {
            console.log(e);
        }
    };
};

export const startRegisterEmailPasswordName = (email, password, name) => {
    return async (dispatch) => {
        try {
            await firebase
                .auth()
                .createUserWithEmailAndPassword(email, password)
                .then(async ({ user }) => {
                    await user.updateProfile({
                        displayName: name,
                    });

                    dispatch(login(user.uid, user.displayName));
                })
                .catch(({ message }) => {
                    Swal.fire({ title: 'Error', text: message, icon: 'error' });
                    dispatch(setError(message));
                });
        } catch (e) {
            console.log(e);
        }
    };
};

export const startGoogleLogin = () => {
    return async (dispatch) => {
       
            await firebase
                .auth()
                .signInWithPopup(googleAuthProvider)
                .then(({ user }) => {
                    dispatch(login(user.uid, user.displayName));
                })
                .catch((error) => {
                    console.log(error);
                });
         
    };
};

export const login = (uid, displayName) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName,
        },
    };
};

export const startLogout = () => {
    return (dispatch) => {
        firebase.auth().signOut();

        dispatch(logout());
        dispatch(noteLogout());
    };
};

export const logout = () => ({
    type: types.logout,
});
