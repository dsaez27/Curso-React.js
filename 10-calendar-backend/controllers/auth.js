const User = require('../models/User');
const bcrypt = require('bcryptjs');
const { generateJwt } = require('../helpers/jwt');

const crearUsuario = async (req, res) => {
    const { email, password } = req.body;

    try {
        let user = await User.findOne({ email });
        console.log(user);
        if (user) {
            return res.status(400).json({
                ok: false,
                msg: 'User already exists',
            });
        }

        user = new User(req.body);

        //Encriptar contraseÃ±a
        const salt = await bcrypt.genSalt(10);
        user.password = await bcrypt.hash(password, salt);

        await user.save();

        //Generate token
        const token = await generateJwt(user._id, user.name);

        res.status(201).json({
            ok: true,
            uid: user._id,
            name: user.name,
            token,
        });
    } catch (error) {
        console.log(error);
        res.status(500).json({
            ok: false,
            msg: 'Plase report this error to the admin',
        });
    }

    res.status(201).json({
        ok: true,
        msg: 'User created',
    });
};

const loginUsuario = async (req, res) => {
    const { email, password } = req.body;

    try {
        let user = await User.findOne({ email });
        console.log(user);
        if (!user) {
            return res.status(400).json({
                ok: false,
                msg: 'User not exists with this email',
            });
        }

        //Validate password
        const validPassword = await bcrypt.compare(password, user.password);

        if (!validPassword) {
            return res.status(400).json({
                ok: false,
                msg: 'Invalid password',
            });
        }

        //Generate token
        const token = await generateJwt(user._id, user.name);

        res.status(200).json({
            ok: true,
            uid: user._id,
            name: user.name,
            token,
        });
    } catch (error) {
        console.log(error);
        res.status(500).json({
            ok: false,
            msg: 'Plase report this error to the admin',
        });
    }
};

const revalidarToken = async (req, res) => {
    const uid = req.uid;
    const name = req.name;

     //Generate token
     const token = await generateJwt(uid, name);

    res.json({
        ok: true,
        token
    });
};

module.exports = {
    crearUsuario,
    loginUsuario,
    revalidarToken,
};
