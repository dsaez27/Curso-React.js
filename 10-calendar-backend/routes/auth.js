const { Router } = require('express');
const {
    crearUsuario,
    loginUsuario,
    revalidarToken,
} = require('../controllers/auth');
const router = Router();
const { check } = require('express-validator');
const { validarCampos } = require('../middlewares/validarCampos');
const { validarJwt } = require('../middlewares/validarJwt');

router.post(
    '/new',
    [
        check('name', 'Name is required').not().isEmpty(),
        check('email', 'Email is required').isEmail(),
        check('password', 'Password to be at least 6 characters').isLength({
            min: 6,
        }),
        validarCampos,
    ],
    crearUsuario
);
router.post(
    '/',
    [
        check('email', 'Email is required').isEmail(),
        check('password', 'Password to be at least 6 characters')
            .not()
            .isEmpty(),
        validarCampos,
    ],
    loginUsuario
);
router.get('/renew', validarJwt, revalidarToken);

module.exports = router;
