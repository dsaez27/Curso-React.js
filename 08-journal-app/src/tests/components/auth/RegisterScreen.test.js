import React from 'react';
import { mount } from 'enzyme';
import { RegisterScreen } from './../../../components/auth/RegisterScreen';
import '../../../setupTest';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { MemoryRouter } from 'react-router-dom';
import { types } from './../../../redux/types';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null,
    },
};

let store = mockStore(initState);
//store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <RegisterScreen />
        </MemoryRouter>
    </Provider>
);

describe('Pruebas en RegisterScreen', () => {
    test('Debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de hacer el dispatch de la acciÃ³n respectiva', () => {
        const emailField = wrapper.find('input[name="email"]');
        emailField.simulate('change', { target: { name: 'email', value: '' } });
        wrapper.find('form').simulate('submit', { preventDefault() {} });
        const actions = store.getActions();
        expect(actions[0]).toEqual({
            type: types.uiSetError,
            payload: 'Email is not valid'
        });
    });

    test('Debe de mostrar la caja de alerta con el error', () => {
        const initState = {
            auth: {},
            ui: {
                loading: false,
                msgError: 'El email no es correcto',
            },
        };
        
        let store = mockStore(initState);
        
        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );

        expect(wrapper.find('.auth__alert-error').exists()).toBe(true);
        expect(wrapper.find('.auth__alert-error').text().trim()).toBe(initState.ui.msgError);
    })
    
});
