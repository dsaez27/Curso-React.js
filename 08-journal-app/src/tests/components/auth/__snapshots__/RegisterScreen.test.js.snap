// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en RegisterScreen Debe de mostrarse correctamente 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input mb-1"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Daniel"
    />
    <input
      autoComplete="off"
      className="auth__input mb-1"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="dssh26@outlook.com"
    />
    <input
      autoComplete="off"
      className="auth__input mb-1"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <input
      autoComplete="off"
      className="auth__input mb-1"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm Password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
